src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt

```diff
@@ -537,33 +537,82 @@ internal class ModelCheckingMonitorTracker : MonitorTracker {
 
 class ModelCheckingParkingTracker(val allowSpuriousWakeUps: Boolean = false) : ParkingTracker {
 
-    private val parked = mutableThreadMapOf<Boolean>()
+    /**
+     * Enum representing the possible states of a thread in the parking tracker:
+     *
+     *   - [State.UNPARKED] --- thread is not parked;
+     *   - [State.PARKED] --- thread is parked and waiting;
+     *   - [State.PERMITTED] --- thread has a permit to unpark;
+     *   - [State.INTERRUPTED] --- thread was interrupted.
+     *
+     *     +----------+                     +--------+
+     *     | UNPARKED | --- [ park() ] ---> | PARKED |
+     *     +----------+                     +--------+
+     *     /\      |                          |
+     *     |       |--------------------------|
+     *     |                            |
+     *     |                            | [ unpark() | interrupt() ]
+     *     |                            |
+     *     |                            v
+     *     |                      +-------------------------+
+     *     |--[ waitUnpark() ]--  | PERMITTED / INTERRUPTED |
+     *                            +-------------------------+
+     *
+     */
+    private enum class State {
+        UNPARKED,
+        PARKED,
+        PERMITTED,
+        INTERRUPTED,
+    }
+
+    private class Handle {
+        var state: State = State.UNPARKED
+    }
+
+    private val handle = mutableMapOf<ThreadId, Handle>()
```

Comment:
Consider renaming to `threadStates`, storing `State` instead of `Handle`, and getting rid of the `Handle` abstraction.